transports/quic/src/transport.rs

```diff
@@ -0,0 +1,707 @@
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        self.dialer.remove(&socket_addr.ip().into());
+
+        Ok(listener_id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        if !is_quic_addr(listen) || !is_quic_addr(observed) {
+            return None;
+        }
+        Some(observed.clone())
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| CoreTransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(CoreTransportError::MultiaddrNotSupported(addr));
+        }
+        let mut listeners = self
+            .listeners
+            .iter_mut()
+            .filter(|l| {
+                let listen_addr = l.endpoint_channel.socket_addr();
+                SocketFamily::is_same(&listen_addr.ip(), &socket_addr.ip())
+                    && listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()
+            })
+            .collect::<Vec<_>>();
+
+        // Try to use pick a random listener to use for dialing.
+        let dialing = match listeners.choose_mut(&mut thread_rng()) {
+            Some(listener) => listener.dialer_state.new_dial(socket_addr),
+            None => {
+                // No listener? Get or create an explicit dialer.
+
+                let socket_family = socket_addr.ip().into();
+                let dialer = match self.dialer.entry(socket_family) {
+                    Entry::Occupied(occupied) => occupied.into_mut(),
+                    Entry::Vacant(vacant) => {
+                        vacant.insert(Dialer::new::<P>(self.config.clone(), socket_family)?)
+                    }
+                };
+
+                dialer.state.new_dial(socket_addr)
+            }
+        };
+
+        Ok(async move {
+            let connection = dialing.await??;
+            let final_connec = Connecting::from_connection(connection).await?;
+            Ok(final_connec)
+        }
+        .boxed())
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<Self::Dial, CoreTransportError<Self::Error>> {
+        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
+        // `addr`. See DCUtR specification below.
+        //
+        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
+        self.dial(addr)
```

Comment:
So this isn't implemented at all then? Since this isn't having th enode act as a listener. Maybe we should at least log a warning, if not return an error.

Reply:
Changed it in fdca477c381a71cd8c52be341c3f74bdc8fd0198 to return an error here.